apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RemoveMe/ResumableUpload.cs

```diff
@@ -0,0 +1,1077 @@
+﻿/*
+Copyright 2012 Google Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using System;
+using System.Collections;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Reflection;
+
+using Google.Apis.Http;
+using Google.Apis.Logging;
+using Google.Apis.Media;
+using Google.Apis.Requests;
+using Google.Apis.Services;
+using Google.Apis.Testing;
+using Google.Apis.Util;
+using Google.Apis.Json;
+
+namespace Google.Apis.Upload
+{
+    public sealed class ResumableUploadSessionOptions
+    {
+        public HttpClient HttpClient { get; set; }
+
+        internal ConfigurableHttpClient ConfigurableHttpClient { get { return HttpClient as ConfigurableHttpClient; } }
+
+        public Action<HttpRequestMessage> ModifySessionUriRequest { get; set; }
+
+        public ISerializer Serializer { get; set; }
+
+        public string ServiceName { get; set; }
+    }
+
+    public class TmpResumableUpload
+    {
+        protected ConfigurableHttpClient HttpClient { get; }
+
+        private ResumableUploadSessionOptions Options { get; }
+
+        private string SignedUrl { get; set; }
```

Comment:
Oh right, this concept should only exist in the storage library. To answer your question, the signed URL is different from the session URI, but it sounds like maybe you're asking if the session URI resulting from the POST of a signed URL is different from any other session URI. If that is the case, then no, as far as I understand it, there is no difference.

So just to make sure I understand this suggestion, what class would be abstract here (and therefore have `GetSessionUriAsync` defined on it)? `(Tmp)ResumableUpload` or the `ResumableUploadSessionOptions`? If its the former, I'm a little concern that this might tie things down a bit too much: `ResumableUpload<TRequest>` would probably need to be the derived class which overrides `GetSessionUriAsync` to use the 'standard' way to create the session, so anything which needs one of the generic resumable upload classes would be locked into using this mechanism. Doing something on the `ResumableUploadSessionOptions` might work though (although it may make sense to rename if it it's going to have a class hierarchy and/or be more than a simple options holder). I gave it a callback so callers could add headers to the create session request message, so it is already providing some customization around session creation. Maybe we just need to go a bit further and make it fully responsible for session creation. Or we could leave it as an options class and have one option be something like a `ResumableUploadSessionCreator` instance which has the derived classes as you described. If null, we default to using the standard mechanism.